packages/sit-onyx/src/i18n/index.ts

```diff
@@ -0,0 +1,135 @@
+   *
+   * @example
+   * ```ts
+   * import deDE from "sit-onyx/locales/de-DE.json";
+   * {
+   *   messages: {
+   *     // English is always supported so we don't need to add it here
+   *     'de-DE': deDE
+   *   }
+   * }
+   * ```
+   */
+  messages?: Record<string, DeepPartial<Translation> | Record<string, TranslationValue>>;
+};
+
+const I18N_INJECTION_KEY = Symbol() as InjectionKey<ReturnType<typeof createI18n>>;
+
+/**
+ * Creates a new i18n instance.
+ */
+const createI18n = (options?: ProvideI18nOptions) => {
+  /**
+   * Current locale.
+   * @default "en-US"
+   */
+  const locale = computed(() => unref(options?.locale) ?? "en-US");
+
+  const messages = computed(() => {
+    if (options?.messages && locale.value in options.messages) {
+      return options.messages[locale.value];
+    }
+    return enUS;
+  });
+
+  /**
+   * Gets the translation for the given key.
+   * If message is not found for current locale, English fallback will be used.
+   * @param placeholders Placeholders that will be replaced in the message string
+   * For pluralization, you must provide the placeholder `n`.
+   */
+  const t = computed(() => {
+    return (
+      key: ObjectToDottedStrings<Translation>,
+      placeholders: Record<string, string | number | undefined> = {},
+    ): string => {
+      // use English message as fallback
+      let message = resolveMessage(key, messages.value) ?? resolveMessage(key, enUS) ?? "";
+
+      if (placeholders) {
+        Object.entries(placeholders).forEach(([key, value]) => {
+          if (value === undefined) return;
+          // "gi" is used to replace all occurrences because String.replaceAll() is not available
+          // in our specified EcmaScript target
+          message = message.replace(new RegExp(`{${key}}`, "gi"), value.toString());
+        });
+      }
+
+      const pluralizationValue = typeof placeholders.n === "number" ? placeholders.n : undefined;
+      return resolvePluralization(message, pluralizationValue);
+    };
+  });
+
+  return { locale, t };
+};
+
+/**
+ * Provides a global i18n instance that is used by Onyx.
+ * Must only be called once in the `App.vue` file.
+ */
+export const provideI18n = (options: ProvideI18nOptions) => {
+  provide(I18N_INJECTION_KEY, createI18n(options));
+};
+
+/**
+ * Injects the Onyx i18n instance.
+ */
+export const injectI18n = () => {
+  const fallbackValue = createI18n();
+  return inject(I18N_INJECTION_KEY, fallbackValue);
+};
+
+/**
+ * Resolves the given dotted key (e.g. `a.b.c`) to the translation value of the given messages.
+ * @returns Message value or undefined if translation does not exist.
+ */
+const resolveMessage = (
+  key: ObjectToDottedStrings<Translation>,
+  messages: DeepPartial<Translation>,
+): string | undefined => {
+  // see https://stackoverflow.com/a/6394168
+  const message = key.split(".").reduce<TranslationValue | undefined>((obj, i) => {
+    if (!obj || typeof obj === "string") return obj;
```

Comment:
I tried to dig deeper to understand this better, but could never enter this condition when trying it with our "required" test input.. how would this be possible?

```
  console.log("## 1 params", key, messages);
  console.log("## 2 split", key.split("."));
  // see https://stackoverflow.com/a/6394168
  const message = key.split(".").reduce<TranslationValue | undefined>((obj, i) => {
    console.log("## 3 red iteration with", i, "obj is", obj);
    if (!obj || typeof obj === "string") {
      console.log("## 4 result is", obj);
      return obj;
    }
    console.log("## 5 iteration result is", obj[i]);
    return obj[i];
  }, messages);

  console.log("## 6 original messages is", messages);
  return message && typeof message === "string" ? message : undefined;
```
Example: ## 4 was never logged

(fyi, imho a unit test for this function might be helpful and also maybe moving it to a util because this is not just for translations but to traverse into a given object by the given key path)